### DISCUSSION TITLE
RTC getEpoch race condition

### DISCUSSION THREAD

**Discussion by ant32t** [stm32duino/Arduino_Core_STM32#2600] [General]
Found a race condition for STM32RTC getEpoch

https://github.com/stm32duino/STM32RTC/blob/a3036e0e9605ddc0e13c7cc259082bfe394b6e77/src/STM32RTC.cpp#L1110
```cpp
time_t STM32RTC::getEpoch(uint32_t *subSeconds)
{
  struct tm tm;

  syncDate();
  syncTime(); // <-- date may have changed and time sometimes reads 00:00:00 of the following date.
  ```

**Comment by ant32t** [comment]
a fix might just check if the hour is 0 ( maybe minute and second too) and call syncDate again

**Reply by fpistm** [reply]
Hi @ant32t 
Seems coherent. Please, could you open an issue on STM32RTC repo and provide a PR for this?
I guess testing the 3 parts (hours, minutes and seconds) is a good idea.

**Reply by fpistm** [reply]
Hi @ant32t 
did you plan to open an issue and create a PR ?

**Reply by ant32** [reply]
@fpistm I was planning on but forgot. Did it now. Please let me know if I need to do something different.

https://github.com/stm32duino/STM32RTC/issues/114
https://github.com/stm32duino/STM32RTC/pull/115

**Reply by uzi18** [reply]
@ant32 why not just reverse - simply sync date after time

**Reply by ant32** [reply]
@uzi18 The race condition stays but is inverted. In that case it might read time as 23:59:59 but date as the following day meaning it would be ahead by 24 hours of actual time.

**Reply by uzi18** [reply]
@ant32 right, so maybe this check should be in syncTime() function

**Reply by ant32** [reply]
@uzi18 I'm not sure how that would work. From what I can see the other places where syncTime is used only the time is important and other places where syncDate is called only the date is important.

Actually it is also used in the begin() function which could result in a race condition there possibly? It doesn't look to me like something that needs an exact date/time. @fpistm might know if it racing there is an issue.

**Reply by fpistm** [reply]
Hi @ant32

Thanks for the issue and PR.
I guess the begin should also be updated as well and order also like in `getEpoch` to be aligned.

**Reply by ant32** [reply]
@fpistm added the check to the begin function and changed order to match

**Reply by fpistm** [reply]
Thanks @ant32t 
It is merged.
